kvbc/src/categorization/kv_blockchain.cpp

```diff
@@ -470,13 +473,58 @@ void KeyValueBlockchain::deleteGenesisBlock() {
 
   // Iterate over groups and call corresponding deleteGenesisBlock,
   // Each group is responsible to fill its deltetes to the batch
+  const auto start = std::chrono::steady_clock::now();
+  std::vector<std::future<void>> futures;
+  futures.reserve((*block).data.categories_updates_info.size());
+  std::vector<detail::LocalWriteBatch> write_batches;
+  write_batches.reserve((*block).data.categories_updates_info.size());
   for (auto&& [category_id, update_info] : (*block).data.categories_updates_info) {
-    std::visit([&genesis_id, category_id = category_id, &write_batch, this](
-                   const auto& update_info) { deleteGenesisBlock(genesis_id, category_id, update_info, write_batch); },
-               update_info);
+    uint64_t num_of_keys = 0;
+    // Decide whether to perform the deletion on a seperate thread based on the number of keys
+    // Immutable category, should always be sequnetial as its deletion is very fast.
+    if (std::holds_alternative<VersionedOutput>(update_info)) {
+      num_of_keys = std::get<VersionedOutput>(update_info).keys.size();
+    } else if (std::holds_alternative<BlockMerkleOutput>(update_info)) {
+      num_of_keys = std::get<BlockMerkleOutput>(update_info).keys.size();
+    }
+    std::visit(
+        [&genesis_id, category_id = category_id, &write_batches, &futures, &num_of_keys, this](
+            const auto& update_info) {
+          write_batches.push_back(detail::LocalWriteBatch());
+          if (num_of_keys > concurrent_threshold) {
+            futures.push_back(prunning_thread_pool_.async(
+                [&](BlockId genesis_id,
+                    std::string category_id,
+                    const auto& update_info,
+                    detail::LocalWriteBatch& write_batch) {
+                  LOG_DEBUG(CAT_BLOCK_LOG, "Deletion of " << category_id << " will be performed in a seperate thread");
+                  deleteGenesisBlock(genesis_id, std::ref(category_id), std::ref(update_info), std::ref(write_batch));
+                },
+                genesis_id,
+                category_id,
+                std::ref(update_info),
+                std::ref(write_batches.back())));
+
+          } else {
+            deleteGenesisBlock(genesis_id, category_id, update_info, write_batches.back());
+          }
+        },
+        update_info);
+  }
+  // wait for jobs to finish
+  for (std::vector<std::future<void>>::size_type i = 0; i < futures.size(); ++i) {
```

Comment:
I wonder, wouldn't it be better not to wait for the actual delete?
Instead, we can simply change the current genesis block id (increase it by one) and let the actual delete happen in its own time.
It is a bit risky, but maybe it is worth a try, WDYT?

Reply:
I don't think it can work with our current implementation.
For example, if we delete a range, it will allow deleting more than one block in parallel, and categories share data across blocks e.g., the Merkle active keys column family.